### ISSUE TITLE
sanity check with lingua

### ISSUE THREAD

**Issue by SeunghyunSEO** [facebookresearch/optimizers#50]
![image](https://github.com/user-attachments/assets/4cfe7cf0-cd51-411b-82e6-e6c6b7c99b0a)

i compared shampoo vs adamw with lingua, and here is [my training details and observation notes](https://github.com/SeunghyunSEO/lingua/tree/main/exp_logs#shampoo-vibe-check)

TL;DR

- it does not show a significant gap compared to adamw baseline in small scale (small batch, small training tokens) 
- MFU of 1 node with fsdp2 (8,1 no shard mesh) is horrible

i just need your advice! is it an acceptable gap? and any recommendation for better MFU?
it's my first time to use shampoo, so i should optimize MFU from now but i just want to know what is proper MFU

**Comment by SeunghyunSEO** [comment]
> Hi @SeunghyunSEO , 
> 
> 
> 
> Thanks for providing feedback, and I have following quests:
> 
> 1. I check your [training details and observation notes](https://github.com/SeunghyunSEO/lingua/tree/main/exp_logs#shampoo-vibe-check), and could not find the `beta1` you are using, do you know that value (i.e., what is the value of `args.beta1`)? Ideally it should be identical to your AdamW setting.
> 
> 2. You might consider using the newly released SOAP/eigenvalue-corrected Shampoo, here I link [@runame 's recommended setting for people who were using Adam before](https://x.com/runame_/status/1854242161857835020/photo/1); for AdamW, you might need to set `use_decoupled_weight_decay=True` as well. This version of Shampoo is very effective on models who got good training results on Adam/AdamW.
> 
> 
> 
> I am not familiar with lingua personally but hope this will help. Again, thanks for providing this feedback, and please do let us know how your experiments go (either good of bad).

i love how you response so kindly and fast!
beta1 is 0.9 and beta2 is 0.999 for shampoo and 0.95 for adam grafting.
for decoupled wd flag of this module, you mean unless torch default adamw use 0.1~0.01 because it is mupltipled by current lr (peaklr*decay_factor), it use truly independent wd right?
but i dont get it the comment, 'for AdamW, you might need to set `use_decoupled_weight_decay=True` as well.'
you mean i should set shampoo config as true? or adamw config? or soap??
if former one, i already set shampoo config's decoupled wd as true and wd as 0.00001 (in my first attempt, i use 0.1 and use decoupld wd flag false but it's not work as i expected because it does not follow pytorch's default behavior and i realize you guys impl decoupled wd by own)

**Comment by SeunghyunSEO** [comment]
@tsunghsienlee sry for the late reply, ty so much I'll try soap and some combination betas and eps!
but may i ask you to advice about MFU too?

and i want you to notice one more thing! 
here's [my toy example for distributed shampoo test](https://github.com/SeunghyunSEO/lingua/tree/main/exp_srcs#shampoo)
when i use 4gpus and test fsdp1/hsdp/fsdp2, i got significant performance degradation with fsdp2 dp_shard > 1.
idk it can cause potential issue when i scale up shampoo with hybrid fsdp2 (like mics).
you can check my [shampoo test code here](https://github.com/SeunghyunSEO/lingua/blob/main/exp_srcs/test_shampoo.py)

**Comment by runame** [comment]
Great to see you trying Shampoo with lingua, I also wanted to do this at some point!

Regarding your two concerns:
- I'm not sure how big of a gap we should expect in this setting. Sometimes already small improvements in convergence speed/final loss are hard to achieve and then the gap you observe could already be significant. I would also try SOAP by setting `preconditioner_computation_config=QREigenvalueCorrectionConfig()`, maybe with `precondition_frequency=10` (increase it if it is too slow). Remember to not use grafting in this setting.
- I need to learn more about FSDP2 before I can help with this, but I will ask for support from other people.

**Comment by hjmshi** [comment]
Hi @SeunghyunSEO - Thanks for your interest in our work! 

Just to provide some more clarification, we support both FSDP and FullyShard (FSDP2). The way we support this, however, is by respecting the parameter sharding that those distributed training frameworks give to us. In the FSDP case, we will re-shape the flattened parameters into blocks of the original tensor; in FSDP2 / FullyShard, we will only apply Shampoo directly to the rows that Shampoo receives from FullyShard. 

Because of this, you should expect that convergence will worsen, especially when parameters are oversharded. You can interpret this as further blocking the parameters beyond the max_preconditioner_dim or block size we determine. Other techniques like tensor parallelism can also contribute to this effect. In our experience, we haven't seen worse convergence with FSDP since it flattens, concatenates, and chunks all of the parameters within each FSDP module, but FullyShard can over-shard much more easily since sharding occurs on each parameter. This may play into some of the discrepancies you're observing.

If it's possible, it'd be great to start with a small-scale DDP setup, and tune to make sure you can achieve better convergence.

Hope this helps!
